independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java

```diff
@@ -164,30 +161,23 @@ public synchronized QuarkusClassLoader getAugmentClassLoader() {
             //we want a class loader that can load the deployment artifacts and all their dependencies, but not
             //any of the runtime artifacts, or user classes
             //this will load any deployment artifacts from the parent CL if they are present
-            Set<AppArtifact> deploymentArtifacts = new HashSet<>();
             for (AppDependency i : appModel.getFullDeploymentDeps()) {
-                AppArtifactKey key = getKey(i);
-                deploymentArtifacts.add(i.getArtifact());
-                ClassPathElement element = getElement(i.getArtifact());
-                builder.addElement(element);
-                if (appModel.getParentFirstArtifacts().contains(key)) {
-                    //we always load this from the parent if it is available, as this acts as a bridge between the running
-                    //app and the dev mode code
-                    builder.addParentFirstElement(element);
-                }
-            }
-            for (AppDependency userDep : appModel.getUserDependencies()) {
-                if (!deploymentArtifacts.contains(userDep.getArtifact())) {
-                    AppArtifactKey key = getKey(userDep);
-                    ClassPathElement element = getElement(userDep.getArtifact());
-                    if (appModel.getParentFirstArtifacts().contains(key)) {
-                        //this mostly happens when building quarkus itself
+                processCpElement(i.getArtifact(), element -> {
+                    builder.addElement(element);
+                    if (appModel.getParentFirstArtifacts().contains(getKey(i))) {
+                        //we always load this from the parent if it is available, as this acts as a bridge between the running
+                        //app and the dev mode code
                         builder.addParentFirstElement(element);
                     }
-                    builder.addElement(element);
-                }
+                });
             }
 
+            appModel.getAppArtifact().getPaths().forEach(p -> {
+                if (Files.exists(p)) {
+                    builder.addElement(ClassPathElement.fromPath(p));
+                }
+            });
```

Comment:
@stuartwdouglas please check this one. I am adding the root application artifact here. I think it's ok because all the deps are added here besides the root artifact. Full deployment deps above already contain the "user deps".

Reply:
This does not look right, this should be in the deployment CL instead.

In dev mode this CL is persistent, so if you include parts of the application then any changes will not be visible.

Reply:
What about the local projects that appear as "user deps" and among the full deployment dependencies?

Reply:
BTW, there would be test failures if I added it in the deployment CL.

Reply:
Local projects that are extension dependencies and this not hot reloadable are fine, but we can't have anything hot reloadable here.

Reply:
Isn't it happening here https://github.com/quarkusio/quarkus/blob/master/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java#L172 ? Full deployment deps include all the user configured runtime deps including the local hot reloadable projects, afaiu.

Reply:
Hmm, looks like this is an existing problem.

In practice it should be fine as the relevant classes should always be loaded from the runtime/deployment class loader.